api/v1alpha1/types_etcd.go

```diff
@@ -485,3 +567,45 @@ func (e *Etcd) GetRoleName() string {
 func (e *Etcd) GetRoleBindingName() string {
 	return fmt.Sprintf("%s:etcd:%s", GroupVersion.Group, e.Name)
 }
+
+// IsBackupStoreEnabled returns true if backup store has been enabled for the Etcd resource, else returns false.
+func (e *Etcd) IsBackupStoreEnabled() bool {
+	return e.Spec.Backup.Store != nil
+}
+
+// IsMarkedForDeletion returns true if a deletion timestamp has been set and false otherwise.
+func (e *Etcd) IsMarkedForDeletion() bool {
+	return !e.DeletionTimestamp.IsZero()
+}
+
+// GetSuspendEtcdSpecReconcileAnnotationKey gets the annotation key set on an Etcd resource signalling the intent
+// to suspend spec reconciliation for this Etcd resource. If no annotation is set then it will return nil.
+func (e *Etcd) GetSuspendEtcdSpecReconcileAnnotationKey() *string {
+	if metav1.HasAnnotation(e.ObjectMeta, SuspendEtcdSpecReconcileAnnotation) {
+		return pointer.String(SuspendEtcdSpecReconcileAnnotation)
+	}
+	if metav1.HasAnnotation(e.ObjectMeta, IgnoreReconciliationAnnotation) {
+		return pointer.String(IgnoreReconciliationAnnotation)
+	}
+	return nil
+}
+
+// IsReconciliationSuspended returns true if the Etcd resource has the annotation set to suspend spec reconciliation,
+// else returns false.
+func (e *Etcd) IsReconciliationSuspended() bool {
+	suspendReconcileAnnotKey := e.GetSuspendEtcdSpecReconcileAnnotationKey()
+	return suspendReconcileAnnotKey != nil && metav1.HasAnnotation(e.ObjectMeta, *suspendReconcileAnnotKey)
+}
+
+// AreManagedResourcesProtected returns true if the Etcd resource has the disable-resource-protection annotation set,
+// else returns false
+func (e *Etcd) AreManagedResourcesProtected() bool {
+	return !metav1.HasAnnotation(e.ObjectMeta, DisableResourceProtectionAnnotation)
+}
+
+// IsReconciliationInProgress returns true if the Etcd resource is currently being reconciled, else returns false.
+func (e *Etcd) IsReconciliationInProgress() bool {
+	return e.Status.LastOperation != nil &&
```

Comment:
When a etcd spec reconciliation starts then the first thing it does it updates the `LastOperation`. So technically there is a very small delay in updating the `LastOperation` and if `IsReconciliationInProgress` called before it gets updated then it will result in a semantically incorrect result. 

However, consider the case that you have mentioned above: If the etcd spec is reconciled for the very first time, then there are no managed etcd resources (e.g. sts, role, role-binding etc.). Therefore the webhook which actually uses `IsReconciliationInProgress` is not in play yet (unless ofcourse someone manually creates one of the managed resources by hand and sets the `OwnerReference` which usually no one will do). So in this case ideally this behavior should not cause any issues.

For the other case where this is not the first reconciliation, there the `LastOperation` will always be present.

Reply:
Got it. Thanks for explaining